packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol

```diff
@@ -0,0 +1,285 @@
+    // can be created either within a window (last claimants within a window don't get their claim)
+    // or across windows (claimants from window T take rewards from window T+1).
+    function setWindowMerkleRoot(
+        uint256 totalRewardsDistributed,
+        uint256 windowStart,
+        uint256 windowEnd,
+        address rewardToken,
+        bytes32 merkleRoot
+    ) external nonReentrant() onlyOwner {
+        uint256 indexToSeed = lastSeededIndex;
+        lastSeededIndex = indexToSeed.add(1);
+
+        _seedWindow(indexToSeed, totalRewardsDistributed, windowStart, windowEnd, rewardToken, merkleRoot);
+    }
+
+    // Flip window lock to opposite of current value. When `window.locked` is true,
+    // claims are blocked for that window.
+    function toggleWindowLock(uint256 windowIndex) external nonReentrant() onlyOwner {
+        merkleWindows[windowIndex].locked = !merkleWindows[windowIndex].locked;
+    }
+
+    // Overwrite merkle root for specified window index. Owner must deposit `totalRewardsDistributed`
+    // into contract.
+    // TODO: Should we require that `windowIndex <= lastSeededIndex` or should this function just be
+    // a generalized `setWindowMerkleRoot` whereby the owner can set a merkle root for any index?
+    function resetWindowMerkleRoot(
+        uint256 windowIndex,
+        uint256 totalRewardsDistributed,
+        uint256 windowStart,
+        uint256 windowEnd,
+        address rewardToken,
+        bytes32 merkleRoot
+    ) external nonReentrant() onlyOwner {
+        _seedWindow(windowIndex, totalRewardsDistributed, windowStart, windowEnd, rewardToken, merkleRoot);
+    }
+
+    // Emergency methods to transfer rewards into and out of the contract
+    // incase the contract was configured improperly.
+    function withdrawRewards(address rewardCurrency, uint256 amount) external nonReentrant() onlyOwner {
+        IERC20(rewardCurrency).safeTransfer(msg.sender, amount);
+    }
+
+    function depositRewards(address rewardCurrency, uint256 amount) external nonReentrant() onlyOwner {
+        IERC20(rewardCurrency).safeTransferFrom(msg.sender, address(this), amount);
+    }
+
+    /****************************
+     *
+     * Public functions
+     *
+     ****************************/
+
+    // TODO: This method could get pretty gas intensive; is there a way we can reduce the amount
+    // of external `transfer` calls if we precompute all the rewards for each account?
+    // It's a bit tricky because there are multiple reward currencies possible.
+    function claimWindows(Claim[] memory claims) public nonReentrant() {
+        for (uint256 i = 0; i < claims.length; i++) {
+            _claimWindow(claims[i]);
+        }
+    }
+
+    // Claim `amount` of reward tokens for `account`. If `amount` and `account` do not exactly match the values stored
+    // in the merkle proof for this `windowIndex` this method will revert.
+    function claimWindow(Claim memory claim) public nonReentrant() {
```

Comment:
how do you prevent someone from claiming for someone else before vesting is over, is this a problem?

Reply:
You can't prevent it, but I don't see the problems with that. Say person A is owed 100 tokens but wants to wait until they are fully vested to claim all of their tokens, for gas reasons (or tax reasons?). But person B for whatever reason calls `claimWindow` for person A every single second, such that A ends up getting a steady trickle of rewards.

Are there situations where this is a big problem for A?

Reply:
Thanks yea makes a lot of sense this way